### PR TITLE
Add CPU Pinning modal and functionality

### PR DESCRIPTION
**Pull request by glekner**
Add CPU Pinning modal and functionality

### CODE REVIEW

**Comment by glekner** [comment]
---- in Overview -----
Screen Shot 2019-12-19 at 14 33 20

---- Modal -----

Screen Shot 2019-12-19 at 14 33 15

Some screenshots @Ranelim @matthewcarleton . thoughts? 
Also, what should we display in case of no changes to the VM (e.g. the checkbox is unchecked)? just a blank space?

@suomiy @mareklibra @yaacov please review

File: frontend/.vscode/settings.json [hunk 1]
```diff
@@ -34,6 +34,9 @@
     "**/bower_components": true,
     "**/.cache-loader": true,
     "**/public/dist": true
+  },
```
Comment (yaacov) [inline frontend/.vscode/settings.json]: nice, but should it be in this PR ?

Comment (glekner) [reply frontend/.vscode/settings.json]: this was automatically changed by VS Code..everytime i revert the changes it creates the patch again immediately. does it happen to anyone else?

**Comment by matthewcarleton** [comment]
@glekner I'd like @Ranelim to weigh in here but I think this should be called "Resource scheduling" It's more descriptive and I don't think we want to use the term "Device" here. It will conflict with virtual hardware. 
If we go that route then the empty state could be "No resource scheduling applied" or something similar.

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.scss [hunk 2]
```diff
@@ -0,0 +1,8 @@
+.kubevirt-cpu-pinning__container {
```
Comment (matthewcarleton) [inline frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.scss]: Do you need this wrapping div?

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    isOpen,
+    setOpen,
+    title = 'Devices & Resources',
+    handlePromise,
+    inProgress,
+    errorMessage,
+  }: CpuPinningModalProps) => {
+    const [isPinned, setIsPinned] = React.useState<boolean>(isCPUPinned);
+    const [showPatchError, setPatchError] = React.useState<boolean>(false);
+
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: we probably should check the nodes if there are some with `cpumanager=true` label before we start offering the pinning. Or maybe some warning?

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      if (isPinned !== isCPUPinned) {
+        handlePromise(
+          k8sPatch(
+            getVMLikeModel(vmLikeEntity),
+            vmLikeEntity,
+            await getCPUPinningPatch(vmLikeEntity, isPinned),
+          ),
+        )
+          .then(() => setOpen(false))
+          .catch(() => setPatchError(true));
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: why is this necessary? can't we just use the errorMessage?

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    vmLikeEntity,
+    isCPUPinned,
+    isOpen,
+    setOpen,
+    title = 'Devices & Resources',
+    handlePromise,
+    inProgress,
+    errorMessage,
+  }: CpuPinningModalProps) => {
+    const [isPinned, setIsPinned] = React.useState<boolean>(isCPUPinned);
```
[the remark this answers is not in the document]
Comment (matthewcarleton) [reply frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: I agree @suomiy but I think @jelkosz felt like this is more advanced than flavor. @Ranelim have an effort this sprint to look at topology and how it could align with this approach.

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      <Modal
+        title={title}
+        isOpen={isOpen}
+        isSmall
+        onClose={() => setOpen(false)}
+        footer={footer}
+        isFooterLeftAligned
+      >
+        <div className="kubevirt-cpu-pinning__container">
+          <Title className="kubevirt-cpu-pinning__schedule-title" size="sm">
```
Comment (matthewcarleton) [inline frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: It doesn't look like this working properly. I'm seeing an `h1` which we should avoid because we already have an `h1` for the title of the modal.

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+      />
+    );
+
+    return (
+      <Modal
+        title={title}
+        isOpen={isOpen}
+        isSmall
+        onClose={() => setOpen(false)}
+        footer={footer}
```
Comment (matthewcarleton) [inline frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: looks like the alignment of the footer is a bit off. I'm seeing a wrapping div around the footer which shouldn't be needed.
Do we need `footer={footer}`?

File: frontend/.vscode/settings.json (hunk 1 above)
Comment (yaacov) [reply frontend/.vscode/settings.json]: Ref: https://github.com/openshift/console/pull/3803

File: frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts [hunk 8]
```diff
@@ -25,6 +25,8 @@ import { vCPUCount } from './cpu';
 export const getMemory = (vm: VMKind) =>
   _.get(vm, 'spec.template.spec.domain.resources.requests.memory');
 export const getCPU = (vm: VMKind): CPURaw => _.get(vm, 'spec.template.spec.domain.cpu');
+export const getCPUPinningBool = (vm: VMKind) =>
```
[the remark this answers is not in the document]
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts]: How would you name it?

File: frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    CPU.dedicatedCpuPlacement = cpuPinning;
+  }
+
+  const patches = [
+    new PatchBuilder('/spec/template/spec/domain/cpu')
+      .setOperation(PatchOperation.REPLACE)
+      .setValue(CPU)
+      .build(),
+  ];
+  return getVMLikePatches(vmLikeEntity, () => patches);
```
[the remark this answers is not in the document]
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts]: `asVM` is for the `getCPU`

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx (hunk 7 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: yes, custom actions and loading animations. ill fix the left padding

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx (hunk 4 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: This is the new modal footer, took example from `boot-order-modal`@122

File: frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts]: i'm going one level deeper, patching just dedicatedCpuPlacement so no need for cpu parsing

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx (hunk 3 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: can you elaborate?

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx (hunk 6 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: is span acceptable here?

Comment (matthewcarleton) [reply frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: I'd stick to using headings. You could use an `h2`.

File: frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx (hunk 3 above)
Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/components/modals/cpu-pinning-modal/cpu-pinning-modal.tsx]: https://kubevirt.io/user-guide/docs/latest/creating-virtual-machines/dedicated-cpu.html#identifying-nodes-with-a-running-cpu-manager

File: frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts]: this will not work if the cpu field is missing

**Comment by glekner** [comment]
@suomiy I fixed your comments.

regarding this section https://kubevirt.io/user-guide/docs/latest/creating-virtual-machines/dedicated-cpu.html#identifying-nodes-with-a-running-cpu-manager

does this label check require us to search inside all nodes for that specific label? wouldn't it hurt performance?

**Comment by suomiy** [comment]
> @suomiy I fixed your comments.

https://github.com/openshift/console/pull/3805/files#r362642812 still needs some attention

> 
> regarding this section https://kubevirt.io/user-guide/docs/latest/creating-virtual-machines/dedicated-cpu.html#identifying-nodes-with-a-running-cpu-manager
> 
> does this label check require us to search inside all nodes for that specific label? wouldn't it hurt performance?

I think it should be okay performance-wise since we would be fetching just the nodes inside the modal. 
I guess it is not a hard requirement but a warning could be nice to have, IMO.

**Comment by glekner** [comment]
@suomiy sorry I was pushing changes to the wrong branch.

File: frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  const isCPUAvailable = !!getCPU(asVM(vmLikeEntity));
+
+  const patches = [
+    new PatchBuilder(
+      isCPUAvailable
+        ? '/spec/template/spec/domain/cpu/dedicatedCpuPlacement'
+        : '/spec/template/spec/domain/cpu',
+    )
+      .setOperation(PatchOperation.REPLACE)
+      .setValue(isCPUAvailable ? dedicatedCpuPlacement : { cores: 1, dedicatedCpuPlacement })
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts]: @suomiy in case cpu is not available, do we even want to give an option to patch `dedicatedCpuPlacement`?

Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts]: I don't see a reason why not. We should just ensure there is no overlap between resources.requests.cpu and domain.cpu and have CPUs in just one place.

**Comment by glekner** [comment]
- Patch now considers `spec.domain.resources.[requests/limits].cpu`
- Added an Alert to the modal with the `cpumanager=true` warning.

Screen Shot 2020-01-15 at 12 07 18
@matthewcarleton  @Ranelim are we ok with this for now?
@suomiy please review.

**Comment by matthewcarleton** [comment]
@glekner is this a warning or info alert? As a warning it feels like I'm being warned that it isn't going to happen, Info seems more accurate if I'm understanding this correctly.
Also, nit: A node with the cpumanager=true label is required. I also wonder if we should highlight the label. @andybraren do you know if in the console anywhere we treat labels differently when they are inline like this?

**Comment by glekner** [comment]
@matthewcarleton changed alert to info
Screen Shot 2020-01-15 at 17 33 18

**Comment by matthewcarleton** [comment]
@glekner ya I think that's probably too much. We may not need anything, but possibly just bolding the text could help. I'd like to follow console's direction here though and not create a new approach.

**Comment by andybraren** [comment]
@glekner @matthewcarleton I asked @bmignano for advice and she suggests we use the same purple-colored badge that's use elsewhere for node labels, seen below. The OpenShift Console doesn't currently use the standard PatternFly badges.

The markup looks straightforward:

```
<span class="tag-item co-text-node">beta.kubernetes.io/os=linux</span>
```

image

**Comment by matthewcarleton** [comment]
@andybraren thanks! That looks better IMO.
After Andy and I have chatted more about this there are some more questions for this. 

> > @suomiy I fixed your comments.
> 
> [#3805 (comment)](https://github.com/openshift/console/pull/3805#discussion_r362642812) still needs some attention
> 
> > regarding this section https://kubevirt.io/user-guide/docs/latest/creating-virtual-machines/dedicated-cpu.html#identifying-nodes-with-a-running-cpu-manager
> > does this label check require us to search inside all nodes for that specific label? wouldn't it hurt performance?
> 
> I think it should be okay performance-wise since we would be fetching just the nodes inside the modal.
> I guess it is not a hard requirement but a warning could be nice to have, IMO.

If I'm understanding what @suomiy is saying we can determine if there is a node available at this point with the right label? If that's true then we only need to alert them when a node with that label isn't available. Maybe we could reword this to be:

```
Scheduling failure
A node with the label cpumanager=true is required for this virtual machine to be scheduled with dedicated resources.
```
and the checkbox could be 
```
Schedule this workload with dedicated resources (guaranteed policy)
```

**Comment by Ranelim** [comment]
I think that the Hint pattern is not working that well here:

- It is not really a hint, as the user is not suggested with anything
- The title of the hint might cause confusion, I'm reading this as "There is currently an attempt to schedule, (and it fails??)"
- In case we need to provide an alert notification when there are no nodes with that label (or if we can provide a pro-active alert), it creates a problem cause the hint pattern is at the top and the alert is at the bottom, which creates a disconnection between them. 

I think we may consider just dropping that label info, and if not, place it as a support text below the checkbox.
Attaching an image of available options, with my comments for each one.

Schedule CPU pinning 00

I also think we should change the modal name as it only reflects workload isolation. We can split it from the CPU & Memory request & limits for pod since it is ok to surface both as independent fields.
This topic will be reevaluated on the "Scheduling area" so it might be better to avoid vague titles now

@matthewcarleton @andybraren @jelkosz @glekner WDYT?

**Comment by glekner** [comment]
Thanks @Ranelim . 
I like Option B, IMO we should show a hint below the checkbox, if the user submits the patch then we can check for available nodes with the label and show the error if none is found.

**Comment by matthewcarleton** [comment]
@glekner @Ranelim I'm voting for option b as well. I don't think there is any issue with how we are using help text with the checkbox, that's doable with a PF help text component from the form. 
A few questions: 
When does the alert fire, when the check the checkbox or when they save? In any case the checkbox would be checked when we see the alert right? 

Do we allow them to save this even if there are no nodes available to be scheduled on?

Should the text in the alert specify that the workload can not be scheduled?

File: frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts [hunk 11]
```diff
@@ -25,6 +25,12 @@ import { vCPUCount } from './cpu';
 export const getMemory = (vm: VMKind) =>
   _.get(vm, 'spec.template.spec.domain.resources.requests.memory');
 export const getCPU = (vm: VMKind): CPURaw => _.get(vm, 'spec.template.spec.domain.cpu');
+export const getResourcesRequestsCPUCount = (vm: VMKind): string =>
+  _.get(vm, 'spec.domain.resources.requests.cpu');
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts]: missing `spec.template` ?

File: frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts [hunk 12]
```diff
@@ -25,6 +25,12 @@ import { vCPUCount } from './cpu';
...
+export const getResourcesLimitsCPUCount = (vm: VMKind): string =>
+  _.get(vm, 'spec.domain.resources.limits.cpu');
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts]: +here

File: frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+  let newCPU;
+  if (resourcesCPU) {
+    newCPU = { dedicatedCpuPlacement };
+  } else {
+    newCPU = { cores: 1, dedicatedCpuPlacement };
+  }
+
+  const patches = [
+    new PatchBuilder(
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts]: can we please branch this to new PatchBuilders to make the intent more clear? and to remove the ternary comparison duplication

**Comment by glekner** [comment]
---- No Nodes Available -----
Screen Shot 2020-01-19 at 18 50 01

---- Node Available ------
Screen Shot 2020-01-19 at 18 50 50

the label is clickable - a link to a search with the needed label.
if no nodes are available the checkbox is still available but you can't submit the modal.
thoughts @Ranelim @matthewcarleton @suomiy?

**Comment by Ranelim** [comment]
Schedule CPU pinning 00
@matthewcarleton @glekner 
So this is a detailed design for option B. Which is our chosen solution.
I've split the image to users with/without nodes permission, making the info that requires permission drop for users without that permission.

I also included a warning hint, before the user clicks the checkbox. I think it is better than having nothing until the user clicks the checkbox.
But I'm not sure we already promote this behavior elsewhere in the console so if this creates a problem then we can just drop the warning.

On a personal note, I think it is a bit weird that the help/secondary text below the checkbox is not actually included in the hint and the hint toggles to warning/alert, but we simply don't do that cause our hints are placed at the top of the page.

Let me know what you think

**Comment by glekner** [comment]
@Ranelim Thanks for your effort. we really want to merge this as a first implementation, most of the warning/error components you mention don't exist yet. we can improve the non-admin and the design flow later, would be nice if you and @matthewcarleton can approve the current design.

Right now the submit button is disabled with the error of no nodes are available, without waiting for interaction with the checkbox.

**Comment by Ranelim** [comment]
> Right now the submit button is disabled with the error of no nodes are available, without waiting for interaction with the checkbox.

This is a product decision, which I cannot take. I discussed this very topic with @jelkosz and He thinks we should allow the user to apply something which results in scheduling failure as it is typical Kubernetes to allow/accept requests which are not yet possible to fulfill. 

I think that it is something we are preventing to do because we already have another solution ready, then we need to change it as long as we can meet our timeline. But it is only my 2cents.

**Comment by Ranelim** [comment]
Updating the design, according to an offline discussion with Gilad and Tomas.
Currently, we drop any notification for a Non-permissions user.
We also drop a warning message pre-checking the checkbox for permission users, and switch the Alert to a warning, after the user has checked the checkbox.

Schedule CPU pinning 00

TBD, determine the modal title with @matthewcarleton

File: frontend/packages/kubevirt-plugin/src/components/modals/resource-scheduling-modal/resource-scheduling-modal.tsx [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+        setOpen(false);
+      }
+    };
+    const footer = (
+      <ModalFooter
+        className=""
+        warningMessage={!loadError && !isNodeAvailable && RESOURCE_NO_NODES_AVAILABLE}
+        errorMessage={showPatchError && errorMessage}
+        inProgress={inProgress || isLoading}
+        isSimpleError={!isNodeAvailable}
```
[the remark this answers is not in the document]
Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/components/modals/resource-scheduling-modal/resource-scheduling-modal.tsx]: IMHO we could do the second one and call useAccessReview before to check the rights for listing nodes

[the remark this answers is not in the document]
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/components/modals/resource-scheduling-modal/resource-scheduling-modal.tsx]: > allow both and the error behaves as warning only
this is the current behaviour.
the error is only for k8s patch rejects

File: frontend/packages/console-shared/src/selectors/common.ts [hunk 15]
```diff
@@ -23,6 +23,10 @@ export const getLabels = <A extends K8sResourceCommon = K8sResourceCommon>(
   value: A,
   defaultValue?: K8sResourceCommon['metadata']['labels'],
 ) => (_.has(value, 'metadata.labels') ? value.metadata.labels : defaultValue);
+export const getLabel = <A extends K8sResourceCommon = K8sResourceCommon>(
+  value: A,
+  label: string,
+) => getLabels(value, {})[label];
```
Comment (suomiy) [inline frontend/packages/console-shared/src/selectors/common.ts]: this function is for the whole code base so we should rather not create a new object on each call

File: frontend/packages/kubevirt-plugin/src/components/modals/dedicated-resources-modal/dedicated-resources-modal.tsx [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+            className="kubevirt-cpu-pinning__checkbox"
+            label="Schedule this workload with dedicated resources (guaranteed policy)"
+            isChecked={isPinned}
+            isDisabled={isLoading}
+            onChange={(flag) => setIsPinned(flag)}
+            id="cpu-pinning-checkbox"
+          />
+          <Text className="kubevirt-cpu-pinning__helper-text" component={TextVariants.small}>
+            Available only on Nodes with labels{' '}
+            <Label kind={NodeModel.kind} name="cpumanager" value="true" expand />
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/modals/dedicated-resources-modal/dedicated-resources-modal.tsx]: this link is pretty cool

File: frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  asVM,
+} from '../../../selectors/vm';
+
+export const getDedicatedCpuPatch = (
+  vmLikeEntity: VMLikeEntityKind,
+  dedicatedCpuPlacement: boolean,
+) => {
+  const vm = asVM(vmLikeEntity);
+  const isCPUAvailable = !!getCPU(vm);
+  const resourcesCPU = getResourcesRequestsCPUCount(vm) || getResourcesLimitsCPUCount(vm);
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts]: we can move this line to the else branch

File: frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts [hunk 18]
```diff
@@ -25,6 +25,12 @@ import { vCPUCount } from './cpu';
 export const getMemory = (vm: VMKind) =>
   _.get(vm, 'spec.template.spec.domain.resources.requests.memory');
 export const getCPU = (vm: VMKind): CPURaw => _.get(vm, 'spec.template.spec.domain.cpu');
+export const getResourcesRequestsCPUCount = (vm: VMKind): string =>
+  _.get(vm, 'spec.template.spec.domain.resources.requests.cpu');
+export const getResourcesLimitsCPUCount = (vm: VMKind): string =>
+  _.get(vm, 'spec.template.spec.domain.resources.limits.cpu');
+export const getDedicatedCpuBool = (vm: VMKind) =>
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts]: I am not much in favor having the type in the function name. What about `isDedicatedCPUPlacement` ?

File: frontend/packages/console-shared/src/selectors/common.ts [hunk 19]
```diff
@@ -23,6 +23,10 @@ export const getLabels = <A extends K8sResourceCommon = K8sResourceCommon>(
   value: A,
   defaultValue?: K8sResourceCommon['metadata']['labels'],
 ) => (_.has(value, 'metadata.labels') ? value.metadata.labels : defaultValue);
+export const getLabel = <A extends K8sResourceCommon = K8sResourceCommon>(
+  value: A,
+  label: string,
+) => _.get(value, ['metadata', 'labels', label]);
```
Comment (suomiy) [inline frontend/packages/console-shared/src/selectors/common.ts]: missing a type

File: frontend/packages/console-shared/src/selectors/common.ts [hunk 20]
```diff
@@ -23,6 +23,10 @@ export const getLabels = <A extends K8sResourceCommon = K8sResourceCommon>(
   value: A,
   defaultValue?: K8sResourceCommon['metadata']['labels'],
 ) => (_.has(value, 'metadata.labels') ? value.metadata.labels : defaultValue);
+export const getLabel = <A extends K8sResourceCommon = K8sResourceCommon>(
+  value: A,
+  label: string,
```
Comment (suomiy) [inline frontend/packages/console-shared/src/selectors/common.ts]: labelKey? or key?

File: frontend/packages/kubevirt-plugin/src/components/modals/modal/modal-footer.tsx [hunk 21]
```diff
@@ -20,13 +20,15 @@ export const ModalErrorMessage: React.FC<ModalErrorMessageProps> = ({ message })
   </Alert>
 );
 
-type ModalSimpleErrorMessageProps = {
+type ModalSimpleMessageProps = {
   message: string;
+  variant?: 'default' | 'danger' | 'success' | 'warning' | 'info';
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/modals/modal/modal-footer.tsx]: can be
```suggestion
  variant?: AlertProps['variant'];
```

File: frontend/packages/kubevirt-plugin/src/components/modals/modal/modal-footer.tsx [hunk 22]
```diff
@@ -55,7 +59,9 @@ type ModalFooterProps = {
 
 export const ModalFooter: React.FC<ModalFooterProps> = ({
   id,
+  className = 'co-m-btn-bar modal-footer kubevirt-create-nic-modal__buttons',
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/modals/modal/modal-footer.tsx]: it would be better to use ```classNames('co-m-btn-bar modal-footer kubevirt-create-nic-modal__buttons', className)``` for the footer and leave className prop empty.

```
<footer className={classNames('co-m-btn-bar modal-footer kubevirt-create-nic-modal__buttons', className)}>
```

File: frontend/packages/kubevirt-plugin/src/components/modals/modal/modal-footer.tsx [hunk 23]
```diff
@@ -55,7 +59,9 @@ type ModalFooterProps = {
...
   errorMessage = null,
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/modals/modal/modal-footer.tsx]: why all the `null` and `false` values as defaults ? you can keep the these props empty and the result will be the same and the function declaration more tidy

File: frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template-resource.tsx [hunk 24]
```diff
@@ -127,6 +138,21 @@ export const VMTemplateDetailsList: React.FC<VMTemplateResourceListProps> = ({
...
+        title={DEDICATED_RESOURCES}
+        idValue={prefixedID(id, 'dedicated-resources')}
+        canEdit
+        onEditClick={() => setDedicatedResourcesModalOpen(true)}
+        editButtonId={prefixedID(id, 'dedicated-resources-edit')}
+      >
+        <DedicatedResourcesModal
+          vmLikeEntity={template}
+          isOpen={isDedicatedResourcesModalOpen}
+          setOpen={setDedicatedResourcesModalOpen}
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template-resource.tsx]: as metioned above, it would be better to pass `onClose` callback
```
const onClose = React.useCallback(() => setDedicatedResourcesModalOpen(false), []);
...
<DedicatedResourcesModal
          vmLikeEntity={template}
          isOpen={isDedicatedResourcesModalOpen}
          onCLose={onClose}



```

File: frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+  getCPU,
+  getResourcesRequestsCPUCount,
+  getResourcesLimitsCPUCount,
+  asVM,
+} from '../../../selectors/vm';
+
+export const getDedicatedCpuPatch = (
+  vmLikeEntity: VMLikeEntityKind,
+  dedicatedCpuPlacement: boolean,
+) => {
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cpu-patches.ts]: its better to add return type to have clear function contract

```suggestion
): Patch[] => {
```

File: frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts [hunk 26]
```diff
@@ -25,6 +25,12 @@ import { vCPUCount } from './cpu';
 export const getMemory = (vm: VMKind) =>
   _.get(vm, 'spec.template.spec.domain.resources.requests.memory');
 export const getCPU = (vm: VMKind): CPURaw => _.get(vm, 'spec.template.spec.domain.cpu');
+export const getResourcesRequestsCPUCount = (vm: VMKind): string =>
+  _.get(vm, 'spec.template.spec.domain.resources.requests.cpu');
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts]: you can use TS 3.7 optional chaining for all `_.get` usages

File: frontend/packages/kubevirt-plugin/src/components/modals/dedicated-resources-modal/dedicated-resources-modal.tsx [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+    };
+    const footer = (
+      <ModalFooter
+        className=""
+        warningMessage={!loadError && !isNodeAvailable && RESOURCE_NO_NODES_AVAILABLE}
+        errorMessage={showPatchError && errorMessage}
+        inProgress={inProgress || isLoading}
+        isSimpleError={!isNodeAvailable}
+        onSubmit={submit}
+        onCancel={() => setOpen(false)}
```
[the remark this answers is not in the document]
Comment (rawagner) [reply frontend/packages/kubevirt-plugin/src/components/modals/dedicated-resources-modal/dedicated-resources-modal.tsx]: actually pass the name should be `close` as defined in `ModalComponentProps`

**Comment by spadgett** [comment]
/hold
the merge queue is blocked
